File: my-leetcode-solutions/validate-binary-search-tree.ts
// My solution to 98 - Validate Binary Search Tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    const orderedArray: number[] = [];

    const helper = (root: TreeNode | null, orderedArray: number[]): number[] => {
          if(root === null) {
            return [];
        }

        helper(root.left, orderedArray);
        
        orderedArray.push(root.val)
        // console.log(orderedArray)
        helper(root.right, orderedArray);

        return orderedArray;
    }
    
   const output = helper(root, orderedArray);

   // check every value in output is unique

   const outputSet = new Set<number>(output);
   if (output.length !== outputSet.size) {
    return false;
   }

   const sortedOutput = [...output].sort((a, b) => a - b)

    const equal = output.every((value, index) => value === sortedOutput[index])

    if (equal) {
        return true;
    } else {
        return false;
    }

};


// Time Complexity:

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null

    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

const rootTestCases = [
    new TreeNode(2, new TreeNode(1), new TreeNode(3)),
    new TreeNode(5, new TreeNode(1), new TreeNode(4, new TreeNode(3), new TreeNode(6))),
    new TreeNode(5, new TreeNode(4), new TreeNode(6, new TreeNode(3), new TreeNode(7))),
];

for (let i = 0; i < rootTestCases.length; i++) {
    const output = isValidBST(rootTestCases[i]);
    console.log(`Output: ${output} for root as ${JSON.stringify(rootTestCases[i])}`);
}

export {};

// depth first search
// for each node, check that current > left and current < right
// if that condition breaks, return false and don't bother continuing
// if you get to the end and it's still true, return true

// But the problem is, you need to ensure that there are no nodes in left subtreees that are bigger than any node in a level above, and no nodes in right subtrees that are smaller than any level above.

File: my-leetcode-solutions-py/validate-binary-search-tree.py
# My solution to 98 - Validate Binary Search Tree

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: TreeNode | None) -> bool:
        pass


# Time Complexity:

root_test_cases = [
    [2, 1, 3],
    [5, 1, 4, None, None, 3, 6],
]

solution = Solution()

for i in range(len(root_test_cases)):
    nodes = [
        TreeNode(value) if value is not None else None
        for value in root_test_cases[i]
    ]
    children = iter(nodes[1:])

    for node in nodes:
        if node is not None:
            node.left = next(children, None)
            node.right = next(children, None)

    root = nodes[0]
    output = solution.isValidBST(root)
    print(f"Output: {output!r} for root as {root_test_cases[i]!r}")
